File: 2023/task1/task1.java
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int N = in.nextInt(); // количество мест
        int M = in.nextInt(); // количество часов
        int[] C = new int[N]; // количество печенек

        int start = 0;
        int end = 0;

        for (int i = 0; i < N; i++) {
            C[i] = in.nextInt();
            end = Math.max(C[i], end);
        }
        getMinimalAmountToEat(C, M, start, end);
    }

    private static void getMinimalAmountToEat(int[] cookiesAmount, int hours, int start, int end) {
        while (start < end) {
            int middle = (start + end) / 2;
            if (isEatable(middle, cookiesAmount, hours)) {
                end = middle;
            } else {
                start = middle + 1;
            }
        }
        if (isEatable(start, cookiesAmount, hours)) {
            System.out.println(start);
        } else {
            System.out.println(0);
        }
    }

    private static boolean isEatable(int perHour, int[] cookiesAmount, int hours) {
        if (perHour == 0) {
            return false;
        }
        int timeToEat = 0;
        for (int amount : cookiesAmount) {
            timeToEat += (amount + perHour - 1) / perHour;
        }
        return timeToEat <= hours;
    }

}

File: 2023/task1/task1.py
import math

# Проверяем, можно ли успеть съесть все печеньки с выбранным количеством
def can_be_eaten(cookies_amounts, per_hour, hours):
    time_to_eat = 0
    for amount in cookies_amounts:
        time_to_eat += math.ceil(amount / per_hour)

    return time_to_eat <= hours


def get_minimal_amount_to_eat(cookies_amounts, hours):
    # Если количество мест с печеньками больше количества часов - Фёдор точно не успеет
    if len(list(filter(lambda amount: amount > 0, cookies_amounts))) > hours:
        return 0

    start = 1
    end = max(cookies_amounts)

    while start < end:
        # Приближаемся к точному значению через бинарный поиск
        middle = start + math.floor((end - start) / 2)

        if can_be_eaten(cookies_amounts, middle, hours):
            end = middle
        else:
            start = middle + 1

    return end


if __name__ == '__main__':
    N, M = [int(x) for x in input().split()]  # Читаем N и M в первой строке
    C = [int(input()) for _ in range(N)]  # Читаем Cn в последующих строках

    print(get_minimal_amount_to_eat(C, M))
